// resources/applications.js
var fn = require('functions'),
	HTTP = require('http-status'),
	ActionSequence = require('ActionSequence'),
	resource = require('resource');

// resource definition

var ApplicationsHandlerSet = resource.DbTableHandlerSet.extend({

	getYearEnd: function () {
		var now = new Date(),
			yearEnd = new Date(now.getFullYear(), 7, 1);	// TODO: allow to customize year end
		if (yearEnd.getTime() < Date.now()) {
			yearEnd.setYear(now.getFullYear() + 1);
		}
		return yearEnd;
	},

// overrides

	create: function (req, res) {
		req.data.expires = this.getYearEnd();
		this.base(req, res);
	},

	update: function (req, res) {
		if (req.application.approved !== null) {
			res.send('Already approved.', HTTP.FORBIDDEN);
		} else {
			req.data.approved = req.data.approved ? new Date() : null;
			this.base(req, res);
		}
	},
	
	load: function (req, id, next) {
		if (id === 'latest') {
			this.getView(req).select({
				where: this.getParentLink(req)
			}, function (err, rows) {
				next(err, rows.length ? rows[rows.length - 1] : null);
			});
		} else {
			this.base(req, id, next);
		}
	}
	
});

var selfOr = function (priv) {
	return function (req, next) {
		var user = req.user || { id: 0 },
			isSelf = (req.access && (req.access.id === user.id));
		next(null, isSelf ? false : priv);
	}
};

var users = module.exports = (new resource.Resource())
	.use(resource.authenticator({
		index: selfOr('viewUsers'),
		create: selfOr('editUser'),
		show: selfOr('viewUsers'),
		update: 'approveApplications',
		destroy: selfOr('editUser')
	}))
	.use(resource.loadChecker('user'))
	.use(resource.loadChecker('application'))
	.use(resource.validator({
		create: {
			school: ['text', 5, 50],
			grade: ['number', 6, 9]
		},
		update: {
			approved: ['bool', true]
		}
	}))
	.setHandlers(new ApplicationsHandlerSet({
		id: 'application',
		table: 'applications',
		parent: {
			field: 'userId',
			id: 'user'
		}
	}))
	.exportHandlers();
