// resource/DbTableHandlerSet.js
var ResourceHandlerSet = require('./ResourceHandlerSet'),
	HTTP = require('http-status');

var DbTableHandlerSet = module.exports = ResourceHandlerSet.extend({

	constructor: function (config) {
		this.base(fn.merge({
			name: null,	// mandatory
			table: null,	// mandatory
			transform: null,
			view: config.table
		}, config));
	},

// handler methods

	index: function (req, res) {
		this.config.view.select({
			transform: this.config.transform
		}, res.finish);
	},

	create: function (req, res) {
		this.config.table.insert(req.data, function (err, result) {
			if (err) {
				res.finish(err);
			} else {
				res.send(req.path + '/' + result.insertId, HTTP.CREATED);
			}
		});
	},

	load: function (req, id, next) {
		this.config.view.selectSingle({
			where: {
				id: id
			},
			transform: this.config.transform
		}, next);
	},

	show: function (req, res) {
		res.send(req[this.config.name]);
	},

	update: function (req, res) {
		this.config.table.update(req.data, {
			id: req[this.config.name].id
		}, function (err) {
			res.finish(err, 'Resource updated.');
		});
	},

	destroy: function (req, res) {
		this.config.table.remove({
			id: req[this.config.name].id
		}, function (err) {
			res.finish(err, 'Resource deleted.');
		});
	}

});
