// resource/DbTableHandlerSet.js
var ResourceHandlerSet = require('./ResourceHandlerSet'),
	HTTP = require('http-status');

var DbTableHandlerSet = module.exports = ResourceHandlerSet.extend({

	constructor: function (config) {
		this.base(fn.merge({
			name: null,	// mandatory
			table: null,	// mandatory
			transform: null,
			view: config.table
		}, config));
	},

	getItem: function (req) {
		return req[this.config.name];
	},

	getTable: function (req) {
		return req.db[this.config.table];
	},

	getView: function (req) {
		return req.db[this.config.view];
	},

// handler methods

	index: function (req, res) {
		this.getView(req).select({
			transform: this.config.transform
		}, res.finish);
	},

	create: function (req, res) {
		this.getTable(req).insert(req.data, function (err, result) {
			if (err) {
				res.finish(err);
			} else {
				res.send(req.path + '/' + result.insertId, HTTP.CREATED);
			}
		});
	},

	load: function (req, id, next) {
		this.getView(req).selectSingle({
			where: {
				id: id
			},
			transform: this.config.transform
		}, next);
	},

	show: function (req, res) {
		res.send(this.getItem(req));
	},

	update: function (req, res) {
		this.getTable(req).update(req.data, {
			id: this.getItem(req).id
		}, function (err) {
			res.finish(err, 'Resource updated.');
		});
	},

	destroy: function (req, res) {
		this.getTable(req).remove({
			id: this.getItem(req).id
		}, function (err) {
			res.finish(err, 'Resource deleted.');
		});
	}

});
