// database/setup/tables/users.js
var DbTableInitializer = require('../DbTableInitializer'),
	fn = require('functions');

var UsersInitializer = module.exports = DbTableInitializer.extend({

	create: function (callback) {
		this.adapter.create({fields: "\
			`id` int(10) unsigned NOT NULL AUTO_INCREMENT,\
			`username` varchar(20) COLLATE utf8_czech_ci NOT NULL,\
			`password` varchar(32) COLLATE utf8_czech_ci NOT NULL,\
			`name` varchar(50) COLLATE utf8_czech_ci NOT NULL,\
			`email` varchar(50) COLLATE utf8_czech_ci NOT NULL,\
			`lastAccess` timestamp NOT NULL DEFAULT CURRENT_TIMESTAMP,\
			`activationCode` varchar(32) COLLATE utf8_czech_ci NOT NULL,\
			`typeId` int(10) unsigned NOT NULL,\
			PRIMARY KEY (`id`)"}, callback);
	},

	populate: function (callback) {
		this.adapter.insertMore([{
			username: 'root',
			password: 'e955ec0c1cbd0e001b7201de8fe71e47',
			name: 'System Root',
			email: '',
			lastAccess: '2000-00-00 00:00:00',
			activationCode: '',
			typeId: 1
		}, {
			username: 'guest',
			password: '084e0343a0486ff05530df6c705c8bb4',
			name: 'Guest',
			typeId: 2
		}], fn.proxy(function (err) {
			if (!fn.handleError(err, 'Cannot populate users table.', callback)) {
				callback(null);
			}
		}, this));
	}

});
