// database/setup/initializeDb.js
var connectToDb = require('../connectToDb')
	util = require('util'),
	fn = require('functions'),
	ActionSequence = require('ActionSequence'),
	initializers = {
		users: require('./tables/users'),
		usertypes: require('./tables/usertypes'),
		v_users: require('./views/users')
	};

var initializeDb = module.exports = function (options, callback) {
	var dbName = options.config.database;
	options.config = fn.merge({}, options.config);
	delete options.config.database;
	
	var connection = connectToDb(options),
		cb = function (err) {
			connection.close(function () {
				callback(err);
			});
		},
		createDb = function (callback) {
			connection.query(util.format("CREATE DATABASE `%s`;", dbName), function (err) {
				if (!fn.handleError(err, 'Cannot create database.', callback)) {
					callback(null);
				}
			});
		},
		initDb = function (callback) {
			connection.query(util.format("USE `%s`;", dbName), function (err) {
				if (err) {
					if (err.code === 'ER_BAD_DB_ERROR') {
						console.log('Creating database `' + dbName + '`...');
						createDb(function () {
							console.log('Database created.');
							connection.useDb(dbName, callback);
						});
					} else {
						fn.handleError(err, 'Cannot select database. ', callback);
					}
				} else {
					connection.useDb(dbName, callback);
				}
			});
		},
		initTables = function (callback) {
			var seq = new ActionSequence();

			fn.each(initializers, function (klass, table) {
				var initializer = new klass({
					connection: connection,
					table: table
				});
				seq.add(function (callback) {
					console.log('Initializing table `' + table + '`...');
					initializer.init(function (err) {
						if (!fn.handleError(err, 'Cannot initialize table.', callback)) {
							console.log('Table initialized.');
							callback(null);
						}
					});
				});
			});
			seq.run(callback);
		};

	console.log('Starting database initialization...');

	initDb(function (err) {
		if (!fn.handleError(err, cb)) {
			initTables(function (err) {
				if (!fn.handleError(err, cb)) {
					console.log('Database initialized.');
					connection.close();
					cb(null);
				}
			});
		}
	});

};
