// resource/handlers/ApplicationsTableHandlers.js
var DbTableHandlers = require('./DbTableHandlers'),
	fn = require('functions'),
	HTTP = require('http-status');

var ApplicationsTableHandlers = module.exports = DbTableHandlers.extend({

	getYearEnd: function () {
		var now = new Date(),
			yearEnd = new Date(now.getFullYear(), 7, 1);	// TODO: allow to customize year end
		if (yearEnd.getTime() < Date.now()) {
			yearEnd.setYear(now.getFullYear() + 1);
		}
		return yearEnd;
	},

// overrides

	create: function (req, res) {
		req.data.expires = this.getYearEnd();
		this.base(req, res);
	},

	show: function (req, res) {
		fn.merge(req.application, fn.pick(req.user, ['username', 'name', 'email']));
		this.base(req, res);
	},

	update: function (req, res) {
		if (req.application.approved !== null) {
			res.send('Already approved.', HTTP.FORBIDDEN);
		} else {
			req.data.approved = req.data.approved ? new Date() : null;
			this.base(req, res);
		}
	},
	
	load: function (req, id, next) {
		if (id === 'latest') {
			this.getView(req).select({
				where: this.getParentLink(req)
			}, function (err, rows) {
				next(err, rows.length ? rows[rows.length - 1] : null);
			});
		} else {
			this.base(req, id, next);
		}
	}
	
});
